#### INITIALIZE ####
setwd(dirname(rstudioapi::getActiveDocumentContext()$path))

#### USER DEFINED VARIABLES ####
library(pegas)
source("../../phylogenetic_tools/functions_sanger.R")
# https://github.com/Ph-IRES/phylogenetic_tools
source("../../phylogenetic_tools/phylogeny_functions.R")
source("../../phylogenetic_tools/multilocus_phylogeny_fuctions.R") # moved df2fasta() and fasta2df to phylogenetic tools

rm(pkg_info, special_packages, bioc_pkgs, cran_packages, cran_pkgs, pkg)

#### SOURCE ####

# the input files for this script were also generated by `ischnura_luta_*_phylogeny_*.R`
# these were run one by one, rather than sourcing from here

# the input files for this script were generated by `wrangle_blast_results_multilocus.R`
# source("wrangle_blast_results_multilocus.R")


#### Concat Fastas, Deduplicate Haplotypes, Align ####

concatFastas(
  inFilePaths = c(
    "../output/multilocus_phylogeny/16s_coi_28s_h3/ischnura_luta_h3_groupisolates.fasta",
    "../output/multilocus_phylogeny/16s_coi_28s_h3/blast_h3_uniq_rmdups_groupisolates.fa"
  ),
  outFilePath = "../output/multilocus_phylogeny/16s_coi_28s_h3/ischnura_luta_h3_groupisolates_blast.fasta"
)

concatFastas(
  inFilePaths = c(
    "../output/multilocus_phylogeny/16s_coi_28s_h3/ischnura_luta_16s_groupisolates.fasta",
    "../output/multilocus_phylogeny/16s_coi_28s_h3/blast_16s_uniq_rmdups_groupisolates.fa"
  ),
  outFilePath = "../output/multilocus_phylogeny/16s_coi_28s_h3/ischnura_luta_16s_groupisolates_blast.fasta"
)

concatFastas(
  inFilePaths = c(
    "../output/multilocus_phylogeny/16s_coi_28s_h3/ischnura_luta_coi_groupisolates.fasta",
    "../output/multilocus_phylogeny/16s_coi_28s_h3/blast_coi_uniq_rmdups_groupisolates.fa"
  ),
  outFilePath = "../output/multilocus_phylogeny/16s_coi_28s_h3/ischnura_luta_coi_groupisolates_blast.fasta"
)

concatFastas(
  inFilePaths = c(
    "../output/multilocus_phylogeny/16s_coi_28s_h3/ischnura_luta_28s_groupisolates.fasta",
    "../output/multilocus_phylogeny/16s_coi_28s_h3/blast_28s_uniq_rmdups_groupisolates.fa"
  ),
  outFilePath = "../output/multilocus_phylogeny/16s_coi_28s_h3/ischnura_luta_28s_groupisolates_blast.fasta"
)

#### Align FASTAS ####
phyDat_h3 <-
  alignFastaFile(
    inFilePath = "../output/multilocus_phylogeny/16s_coi_28s_h3/ischnura_luta_h3_groupisolates_blast.fasta",
    outFilePath = "../output/multilocus_phylogeny/16s_coi_28s_h3/ischnura_luta_h3_groupisolates_blast_aln.fasta"
  ) %>%
  { .[order(names(.))] }

phyDat_16s <-
  alignFastaFile(
    inFilePath = "../output/multilocus_phylogeny/16s_coi_28s_h3/ischnura_luta_16s_groupisolates_blast.fasta",
    outFilePath = "../output/multilocus_phylogeny/16s_coi_28s_h3/ischnura_luta_16s_groupisolates_blast_aln.fasta"
  ) %>%
  { .[order(names(.))] }

phyDat_coi <-
  alignFastaFile(
    inFilePath = "../output/multilocus_phylogeny/16s_coi_28s_h3/ischnura_luta_coi_groupisolates_blast.fasta",
    outFilePath = "../output/multilocus_phylogeny/16s_coi_28s_h3/ischnura_luta_coi_groupisolates_blast_aln.fasta"
  ) %>%
  { .[order(names(.))] }

phyDat_28s <-
  alignFastaFile(
    inFilePath = "../output/multilocus_phylogeny/16s_coi_28s_h3/ischnura_luta_28s_groupisolates_blast.fasta",
    outFilePath = "../output/multilocus_phylogeny/16s_coi_28s_h3/ischnura_luta_28s_groupisolates_blast_aln.fasta"
  ) %>%
  { .[order(names(.))] }


#### MAXIMUM LIKELIHOOD PHYLOGENY x LOCUS ####

tree_opt_bs_16s <-
  phyDat_16s[order(names(phyDat_16s))] %>%
  fasta2tree(
    my_outgroup = "BEA448"
  )

tree_opt_bs_coi <-
  phyDat_coi[order(names(phyDat_coi))] %>%
  fasta2tree(
    my_outgroup = "BEA448"
  )

tree_opt_bs_28s <-
  phyDat_28s[order(names(phyDat_28s))] %>%
  fasta2tree(
    my_outgroup = "BEA448"
  )

tree_opt_bs_h3 <-
  phyDat_h3[order(names(phyDat_h3))] %>%
  fasta2tree(
    my_outgroup = "BEA448"
  )

# save tree as RDS so it could be read in and edited, can read it in with readRDS()
saveRDS(
  tree_opt_bs_16s, 
  file = "../output/multilocus_phylogeny/16s_coi_28s_h3/ischnura_luta_16s_groupisolates_blast_aln_tree.rds"
)

saveRDS(
  tree_opt_bs_coi, 
  file = "../output/multilocus_phylogeny/16s_coi_28s_h3/ischnura_luta_coi_groupisolates_blast_aln_tree.rds"
)

saveRDS(
  tree_opt_bs_28s, 
  file = "../output/multilocus_phylogeny/16s_coi_28s_h3/ischnura_luta_28s_groupisolates_blast_aln_tree.rds"
)

saveRDS(
  tree_opt_bs_h3, 
  file = "../output/multilocus_phylogeny/16s_coi_28s_h3/ischnura_luta_h3_groupisolates_blast_aln_tree.rds"
)

# output the Newick Tree with only the significant bootstrap values
# saveNewickTree(
#   tree_opt_bs$tree %>% 
#     { 
#       .$node.label <- tree_opt_bs$bootstraps_sig
#       .
#     } , 
#   "../output/sanger_curated_ab1_ischnura_luta_H3/tree_ischnura_luta_619_filtered_aligned_2.nwk"
# )

#### Multilocus Concatenated TREE  (not partitioned) ####

p16s <- fasta_16s[ order(names(fasta_16s)) ]
pcoi <- fasta_coi[ order(names(fasta_coi)) ]
p28s <- fasta_28s[ order(names(fasta_28s)) ]
ph3  <- fasta_h3[  order(names(fasta_h3))  ]

# 2) bind into one concatenated phyDat
multi_phy <- cbind.phyDat(p16s, pcoi, p28s, ph3)

# 3) write FASTA directly from the phyDat
write.phyDat(
  multi_phy,
  file   = "../output/multilocus_phylogeny/16s_coi_28s_h3/ischnura_luta_16s-coi-28s-h3_groupisolates_blast_aln_concat.fasta",
  format = "fasta"
)

# collapse to haplotypes
uniqueSeqsFastaFile(
  inFilePath = "../output/multilocus_phylogeny/16s_coi_28s_h3/ischnura_luta_16s-coi-28s-h3_groupisolates_blast_aln_concat.fasta",
  outFilePath = "../output/multilocus_phylogeny/16s_coi_28s_h3/ischnura_luta_16s-coi-28s-h3_groupisolates_blast_aln_concat_haps.fasta"
)

multi_phy_haps <- 
  read.phyDat(
    "../output/multilocus_phylogeny/16s_coi_28s_h3/ischnura_luta_16s-coi-28s-h3_groupisolates_blast_aln_concat_haps.fasta",
    format = "fasta",
    type = "DNA"
  )

# concatenated seq tree
tree_opt_bs_all <-
  multi_phy_haps %>%
  fasta2tree(
    my_outgroup = "BEA157"
  )

saveRDS(
  tree_opt_bs_all, 
  file = "../output/multilocus_phylogeny/16s_coi_28s_h3/ischnura_luta_16s-coi-28s-h3_groupisolates_blast_aln_concat_haps_tree.rds"
)

# output the Newick Tree with only the significant bootstrap values
saveNewickTree(
  tree_opt_bs_all$tree %>% 
    { 
      .$node.label <- tree_opt_bs_all$bootstraps_sig
      .
    } , 
  "../output/multilocus_phylogeny/16s_coi_28s_h3/ischnura_luta_16s-coi-28s-h3_groupisolates_blast_aln_concat_haps_tree.nwk"
)


#### Partitioned MAXIMUM LIKELIHOOD PHYLOGENY x LOCUS 16s COI 28S H3 ####
loci <- c("16s", "coi", "28s", "h3")

models <- 
  c(
    tree_opt_bs_16s$best_model,
    tree_opt_bs_coi$best_model,
    tree_opt_bs_28s$best_model,
    tree_opt_bs_h3$best_model
  )

base_models <- 
  gsub("\\+I.*|\\+G.*", "", models) %>%
  trimws()

# rather than regenerating, use optim.pml from individual loci processing
models_optim.pml <- 
  list(
    tree_opt_bs_16s$model_optim.pml, 
    tree_opt_bs_coi$model_optim.pml, 
    tree_opt_bs_28s$model_optim.pml, 
    tree_opt_bs_h3$model_optim.pml
  )

sp <- 
  pmlPart(
    edge ~ rate + inv + shape + bf + Q,
    # edge ~ rate + inv + shape,
    object = models_optim.pml
    # model  = base_models  # or NULL to inherit
    # control = pml.control(trace = 1)
  )

# use this output to evaluate the formula in pmlPart()
# left side are parameters estimated across all loci as 1
# right side are locus specific parameter estimates.
print(sp)

# trees <- pmlPart2multiPhylo(sp)
# plot(trees, type="phylogram")

best.tree <- sp$fits[[1]]$tree
plot(best.tree, type="phylogram", cex=0.7)

# write.tree(best.tree, file = "partitioned_best_tree.nwk")

# source("bootstrap.pmlPart.R")

# bs_pmlPart <- 
#   bootstrap.pmlPart(
#     x=sp, bs = 11, multicore = TRUE, mc.cores = 11
#   )
# treeBS  <- 
#   plotBS(
#     sp$fits[[1]]$tree, 
#     bs_pmlPart, 
#     type = "phylogram"
#   )

bs_pmlPart <- 
  bootstrap.pmlPart(
    x=sp, bs = 11, multicore = TRUE, mc.cores = 11,
    optEdge  = TRUE,
    optBf    = TRUE,
    optQ     = TRUE,
    optInv   = TRUE,
    optGamma = TRUE,
    optRate  = TRUE,
    optNni   = FALSE
  )

treeBS  <- 
  plotBS(
    sp$fits[[1]]$tree, 
    bs_pmlPart, 
    type = "phylogram"
  )

# CEB, this was commented out, not sure why, if it doesn't work, probably needs minor tweeking
tree_partitioned_bs <-
  plotBS(
    sp$fits[[1]]$tree,
    res,
    # p = threshold_bootstraps/100,
    digits = 2,
    type = "phylogram",
    method = "FBP"
  )
write.tree(tree_partitioned_bs, file = "../output/multilocus_phylogeny/16s_coi_28s_h3/tree_partitioned_bootstraps.nwk")

#### SAVE PARTIIONED DATA TO NEXUS FILE ####
# all_multiPhyDat <- 
#   cbind.phyDat(
#     phyDat_16s,
#     phyDat_coi,
#     phyDat_28s,
#     phyDat_h3
#   )
# # rename genes to remove phyDat_
# idx <- attr(all_multiPhyDat, "index")
# idx$genes <- sub("^phyDat_", "", idx$genes)
# attr(all_multiPhyDat, "index") <- idx
# rm(idx)
# 
# # save sequences to nexus
# file_nexus <- "../output/multilocus_phylogeny/16s_coi_28s_h3/ischnura_luta_16s-coi-28s-h3_partitioned.nex"
# 
# write.nexus.data(
#   as.DNAbin(all_multiPhyDat), 
#   file= file_nexus, 
#   format="dna"
# )
# 
# # create set block to define genes.
# idx <- attr(all_multiPhyDat, "index")
# sites_by_gene <- split(seq_len(nrow(idx)), idx$genes)
# ranges <- function(x) {
#   s <- min(x)
#   e <- max(x)
#   if (s == e) as.character(s) else paste(s, e, sep = "-")
# }
# cat("BEGIN SETS;\n", file = file_nexus, append = TRUE)
# for (gene in names(sites_by_gene)) {
#   rg <- ranges(sites_by_gene[[gene]])
#   cat(sprintf("  Charset %s = %s;\n", gene, rg),
#       file   = file_nexus,
#       append = TRUE)
# }
# cat("END;\n", file = file_nexus, append = TRUE)
# 


